## Imports

In [1]:
import pandas as pd
import numpy as np

pd.options.display.max_rows = 5000
pd.set_option('display.max_columns', None)
pd.options.display.max_seq_items = 2000

## Reading in the datasets

#### Faults dataset

In [2]:
faults = pd.read_csv('../data/J1939Faults.csv',
                     low_memory=False, 
                     parse_dates=[2, 19],
                     infer_datetime_format=True)

In [3]:
faults.head()

RecordID  ESS_Id      EventTimeStamp  \
0         1  990349 2015-02-21 10:47:13   
1         2  990360 2015-02-21 11:34:34   
2         3  990364 2015-02-21 11:35:31   
3         4  990370 2015-02-21 11:35:33   
4         5  990416 2015-02-21 11:39:41   

                          eventDescription  actionDescription  \
0  Low (Severity Low) Engine Coolant Level                NaN   
1                                      NaN                NaN   
2      Incorrect Data Steering Wheel Angle                NaN   
3      Incorrect Data Steering Wheel Angle                NaN   
4                                      NaN                NaN   

                     ecuSoftwareVersion ecuSerialNumber              ecuModel  \
0                               unknown         unknown               unknown   
1                               unknown         unknown               unknown   
2                               unknown         unknown               unknown   
3                               unknown         unknown               unknown   
4  22281684P01*22357957P01*22362082P01*        13063430  0USA13_13_0415_2238A   

   ecuMake  ecuSource   spn  fmi  active  activeTransitionCount  faultValue  \
0  unknown          0   111   17    True                      2         NaN   
1  unknown         11   629   12    True                    127         NaN   
2  unknown         11  1807    2   False                    127         NaN   
3  unknown         11  1807    2    True                    127         NaN   
4    VOLVO          0  4364   17   False                      2         NaN   

  EquipmentID  MCTNumber   Latitude  Longitude   LocationTimeStamp  
0        1439  105354361  38.857638 -84.626851 2015-02-21 11:34:25  
1        1439  105354361  38.857638 -84.626851 2015-02-21 11:35:10  
2        1369  105336226  41.421250 -87.767361 2015-02-21 11:35:26  
3        1369  105336226  41.421018 -87.767361 2015-02-21 11:36:08  
4        1674  105427130  38.416481 -89.442638 2015-02-21 11:39:37

#### Diagnostics dataset

In [4]:
diagnostic = pd.read_csv('../data/VehicleDiagnosticOnboardData.csv')

In [5]:
diagnostic.head()

Id                  Name     Value  FaultId
0   1             IgnStatus     False        1
1   2     EngineOilPressure         0        1
2   3  EngineOilTemperature  96.74375        1
3   4    TurboBoostPressure         0        1
4   5            EngineLoad        11        1

##### Diagnostics Pivot
This data needs to be flattened. We'll use a pivot so it will be ready to join to the Faults data

In [6]:
diagnostic = diagnostic.pivot(index='FaultId', columns='Name', values='Value').reset_index()

In [7]:
diagnostic.head()

Name  FaultId AcceleratorPedal BarometricPressure CruiseControlActive  \
0           1                0              14.21               False   
1           2              NaN                NaN                 NaN   
2           3              NaN                NaN                 NaN   
3           4              NaN                NaN                 NaN   
4           5              NaN                NaN                 NaN   

Name CruiseControlSetSpeed DistanceLtd EngineCoolantTemperature EngineLoad  \
0                 66.48672    423178.7                    100.4         11   
1                      NaN         NaN                      NaN        NaN   
2                      NaN         NaN                      NaN        NaN   
3                      NaN         NaN                      NaN        NaN   
4                      NaN         NaN                      NaN        NaN   

Name EngineOilPressure EngineOilTemperature EngineRpm EngineTimeLtd FuelLevel  \
0                    0             96.74375         0        1632.2      43.2   
1                  NaN                  NaN       NaN           NaN       NaN   
2                  NaN                  NaN       NaN           NaN       NaN   
3                  NaN                  NaN       NaN           NaN       NaN   
4                  NaN                  NaN       NaN           NaN       NaN   

Name          FuelLtd FuelRate FuelTemperature IgnStatus  \
0     12300.907429328        0             NaN     False   
1                 NaN      NaN             NaN      True   
2                 NaN      NaN             NaN       NaN   
3                 NaN      NaN             NaN      True   
4                 NaN      NaN             NaN       NaN   

Name IntakeManifoldTemperature LampStatus ParkingBrake ServiceDistance Speed  \
0                         78.8       1023         True             NaN     0   
1                          NaN       1279          NaN             NaN   NaN   
2                          NaN       1279          NaN             NaN   NaN   
3                          NaN       1279          NaN             NaN   NaN   
4                          NaN      16639          NaN             NaN   NaN   

Name SwitchedBatteryVoltage Throttle TurboBoostPressure  
0                   3276.75      NaN                  0  
1                       NaN      NaN                NaN  
2                       NaN      NaN                NaN  
3                       NaN      NaN                NaN  
4                       NaN      NaN                NaN

## Joining the two datasets

In [8]:
faults_diagnostic = faults.merge(diagnostic, left_on='RecordID', right_on='FaultId')

In [9]:
faults_diagnostic.head()

RecordID  ESS_Id      EventTimeStamp  \
0         1  990349 2015-02-21 10:47:13   
1         2  990360 2015-02-21 11:34:34   
2         3  990364 2015-02-21 11:35:31   
3         4  990370 2015-02-21 11:35:33   
4         5  990416 2015-02-21 11:39:41   

                          eventDescription  actionDescription  \
0  Low (Severity Low) Engine Coolant Level                NaN   
1                                      NaN                NaN   
2      Incorrect Data Steering Wheel Angle                NaN   
3      Incorrect Data Steering Wheel Angle                NaN   
4                                      NaN                NaN   

                     ecuSoftwareVersion ecuSerialNumber              ecuModel  \
0                               unknown         unknown               unknown   
1                               unknown         unknown               unknown   
2                               unknown         unknown               unknown   
3                               unknown         unknown               unknown   
4  22281684P01*22357957P01*22362082P01*        13063430  0USA13_13_0415_2238A   

   ecuMake  ecuSource   spn  fmi  active  activeTransitionCount  faultValue  \
0  unknown          0   111   17    True                      2         NaN   
1  unknown         11   629   12    True                    127         NaN   
2  unknown         11  1807    2   False                    127         NaN   
3  unknown         11  1807    2    True                    127         NaN   
4    VOLVO          0  4364   17   False                      2         NaN   

  EquipmentID  MCTNumber   Latitude  Longitude   LocationTimeStamp  FaultId  \
0        1439  105354361  38.857638 -84.626851 2015-02-21 11:34:25        1   
1        1439  105354361  38.857638 -84.626851 2015-02-21 11:35:10        2   
2        1369  105336226  41.421250 -87.767361 2015-02-21 11:35:26        3   
3        1369  105336226  41.421018 -87.767361 2015-02-21 11:36:08        4   
4        1674  105427130  38.416481 -89.442638 2015-02-21 11:39:37        5   

  AcceleratorPedal BarometricPressure CruiseControlActive  \
0                0              14.21               False   
1              NaN                NaN                 NaN   
2              NaN                NaN                 NaN   
3              NaN                NaN                 NaN   
4              NaN                NaN                 NaN   

  CruiseControlSetSpeed DistanceLtd EngineCoolantTemperature EngineLoad  \
0              66.48672    423178.7                    100.4         11   
1                   NaN         NaN                      NaN        NaN   
2                   NaN         NaN                      NaN        NaN   
3                   NaN         NaN                      NaN        NaN   
4                   NaN         NaN                      NaN        NaN   

  EngineOilPressure EngineOilTemperature EngineRpm EngineTimeLtd FuelLevel  \
0                 0             96.74375         0        1632.2      43.2   
1               NaN                  NaN       NaN           NaN       NaN   
2               NaN                  NaN       NaN           NaN       NaN   
3               NaN                  NaN       NaN           NaN       NaN   
4               NaN                  NaN       NaN           NaN       NaN   

           FuelLtd FuelRate FuelTemperature IgnStatus  \
0  12300.907429328        0             NaN     False   
1              NaN      NaN             NaN      True   
2              NaN      NaN             NaN       NaN   
3              NaN      NaN             NaN      True   
4              NaN      NaN             NaN       NaN   

  IntakeManifoldTemperature LampStatus ParkingBrake ServiceDistance Speed  \
0                      78.8       1023         True             NaN     0   
1                       NaN       1279          NaN             NaN   NaN   
2                       NaN       1279          NaN    

## Cleaning

#### Column names

In [10]:
#My team opted to use snake case for column names

faults_diagnostic = faults_diagnostic.rename(columns={
    "RecordID": "record_id",
    "ESS_Id": "ess_id",
    "EventTimeStamp": "event_timestamp",
    "eventDescription": "event_description",
    "actionDescription": "action_description",
    "ecuSoftwareVersion": "ecu_software_version",
    "ecuSerialNumber": "ecu_serial_number",
    "ecuModel": "ecu_model",
    "ecuMake": "ecu_make",
    "ecuSource": "ecu_source",
    "activeTransitionCount": "active_transition_count",
    "faultValue": "fault_value",
    "EquipmentID": "equipment_id",
    "MCTNumber": "mct_number",
    "Latitude": "latitude",
    "Longitude": "longitude",
    "LocationTimeStamp": "location_timestamp",
    "FaultId": "fault_id",
    "AcceleratorPedal": "accelerator_pedal",
    "BarometricPressure": "barometric_pressure",
    "CruiseControlActive": "cruise_control_active",
    "CruiseControlSetSpeed": "cruise_control_set_speed",
    "DistanceLtd": "distance_ltd",
    "EngineCoolantTemperature": "engine_coolant_temperature",
    "EngineLoad": "engine_load",
    "EngineOilPressure": "engine_oil_pressure",
    "EngineOilTemperature": "engine_oil_temperature",
    "EngineRpm": "engine_rpm",
    "EngineTimeLtd": "engine_time_ltd",
    "FuelLevel": "fuel_level",
    "FuelLtd": "fuel_ltd",
    "FuelRate": "fuel_rate",
    "FuelTemperature": "fuel_temperature",
    "IgnStatus": "ign_status",
    "IntakeManifoldTemperature": "intake_manifold_temperature",
    "LampStatus": "lamp_status",
    "ParkingBrake": "parking_brake",
    "ServiceDistance": "service_distance",
    "Speed": "speed",
    "SwitchedBatteryVoltage": "switched_battery_voltage",
    "Throttle": "throttle",
    "TurboBoostPressure": "turbo_boost_pressure"
})

In [11]:
# The columns 'action_description' and 'fault_value' contain all null values, so they will be dropped

faults_diagnostic = faults_diagnostic.drop(columns=['action_description', 'fault_value'])

In [12]:
faults_diagnostic.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1187335 entries, 0 to 1187334
Data columns (total 43 columns):
 #   Column                       Non-Null Count    Dtype         
---  ------                       --------------    -----         
 0   record_id                    1187335 non-null  int64         
 1   ess_id                       1187335 non-null  int64         
 2   event_timestamp              1187335 non-null  datetime64[ns]
 3   event_description            1126490 non-null  object        
 4   ecu_software_version         891285 non-null   object        
 5   ecu_serial_number            844318 non-null   object        
 6   ecu_model                    1122577 non-null  object        
 7   ecu_make                     1122577 non-null  object        
 8   ecu_source                   1187335 non-null  int64         
 9   spn                          1187335 non-null  int64         
 10  fmi                          1187335 non-null  int64         
 11  active     

In [13]:
faults_diagnostic.head()

record_id  ess_id     event_timestamp  \
0          1  990349 2015-02-21 10:47:13   
1          2  990360 2015-02-21 11:34:34   
2          3  990364 2015-02-21 11:35:31   
3          4  990370 2015-02-21 11:35:33   
4          5  990416 2015-02-21 11:39:41   

                         event_description  \
0  Low (Severity Low) Engine Coolant Level   
1                                      NaN   
2      Incorrect Data Steering Wheel Angle   
3      Incorrect Data Steering Wheel Angle   
4                                      NaN   

                   ecu_software_version ecu_serial_number  \
0                               unknown           unknown   
1                               unknown           unknown   
2                               unknown           unknown   
3                               unknown           unknown   
4  22281684P01*22357957P01*22362082P01*          13063430   

              ecu_model ecu_make  ecu_source   spn  fmi  active  \
0               unknown  unknown           0   111   17    True   
1               unknown  unknown          11   629   12    True   
2               unknown  unknown          11  1807    2   False   
3               unknown  unknown          11  1807    2    True   
4  0USA13_13_0415_2238A    VOLVO           0  4364   17   False   

   active_transition_count equipment_id  mct_number   latitude  longitude  \
0                        2         1439   105354361  38.857638 -84.626851   
1                      127         1439   105354361  38.857638 -84.626851   
2                      127         1369   105336226  41.421250 -87.767361   
3                      127         1369   105336226  41.421018 -87.767361   
4                        2         1674   105427130  38.416481 -89.442638   

   location_timestamp  fault_id accelerator_pedal barometric_pressure  \
0 2015-02-21 11:34:25         1                 0               14.21   
1 2015-02-21 11:35:10         2               NaN                 NaN   
2 2015-02-21 11:35:26         3               NaN                 NaN   
3 2015-02-21 11:36:08         4               NaN                 NaN   
4 2015-02-21 11:39:37         5               NaN                 NaN   

  cruise_control_active cruise_control_set_speed distance_ltd  \
0                 False                 66.48672     423178.7   
1                   NaN                      NaN          NaN   
2                   NaN                      NaN          NaN   
3                   NaN                      NaN          NaN   
4                   NaN                      NaN          NaN   

  engine_coolant_temperature engine_load engine_oil_pressure  \
0                      100.4          11                   0   
1                        NaN         NaN                 NaN   
2                        NaN         NaN                 NaN   
3                        NaN         NaN                 NaN   
4                        NaN         NaN                 NaN   

  engine_oil_temperature engine_rpm engine_time_ltd fuel_level  \
0               96.74375          0          1632.2       43.2   
1                    NaN        NaN             NaN        NaN   
2                    NaN        NaN             NaN        NaN   
3                    NaN        NaN             NaN        NaN   
4                    NaN        NaN             NaN        NaN   

          fuel_ltd fuel_rate fuel_temperature ign_status  \
0  12300.907429328         0              NaN      False   
1              NaN       NaN              NaN       True   
2              NaN       NaN              NaN        NaN   
3              NaN       NaN              NaN       True   
4              NaN       NaN              NaN        NaN   

  intake_manifold_temperature lamp_status parking_brake service_distance  \
0                        78.8        1023          True              NaN   
1                         NaN        1279           NaN              NaN   
2                         N

##### Software Versions

In [14]:
faults_diagnostic[faults_diagnostic['ecu_software_version'].str.contains("?", regex=False, na=False)]

record_id     ess_id     event_timestamp  \
20828        21711    1878624 2015-04-14 11:38:36   
20829        21712    1878625 2015-04-14 11:38:36   
23937        24820    1933193 2015-04-16 19:49:37   
23938        24821    1933194 2015-04-16 19:49:37   
40226        41338    2216854 2015-05-02 14:45:37   
40483        41595    2221486 2015-05-03 09:47:01   
81635        84021    2897660 2015-06-11 07:43:50   
90267        92653    3037645 2015-06-19 06:36:05   
90273        92659    3037705 2015-06-19 06:39:59   
91224        93610    3054166 2015-06-20 01:31:30   
91225        93611    3054179 2015-06-20 01:35:27   
121168      123554    3486482 2015-07-15 13:07:37   
121169      123555    3486493 2015-07-15 13:09:00   
137680      140066    3711771 2015-07-28 04:18:49   
173884      176270    4187612 2015-08-21 10:10:08   
192595      194981    4478021 2015-09-07 04:44:26   
201513      203899    4623295 2015-09-15 07:14:16   
201524      203910    4623396 2015-09-15 07:19:39   
250601      254421    5387234 2015-10-25 04:06:16   
251399      255219    5398340 2015-10-26 06:46:47   
267782      271602    5672679 2015-11-09 13:11:23   
267952      271772    5675840 2015-11-09 15:52:09   
268309      272129    5682031 2015-11-10 03:31:51   
270096      273916    5712857 2015-11-11 10:12:37   
270151      273971    5713816 2015-11-11 10:49:52   
283343      287849    5928171 2015-11-22 11:41:26   
283585      288091    5931105 2015-11-22 15:33:52   
299026      303532    6181884 2015-12-07 13:50:27   
299037      303543    6182259 2015-12-07 14:05:27   
304595      309101    6282207 2015-12-12 02:17:38   
310935      315992    6387858 2015-12-17 14:07:08   
334621      339678    6754717 2016-01-09 09:02:14   
334644      339701    6754906 2016-01-09 09:03:05   
378932      386433    7499008 2016-02-17 03:25:36   
385847      393348    7614898 2016-02-23 07:07:42   
420948      429215    8262650 2016-03-28 00:03:54   
420949      429216    8262652 2016-03-28 00:05:16   
435605      443872    8549139 2016-04-11 13:38:00   
435606      443873    8549140 2016-04-11 13:38:00   
457193      466310    8976744 2016-05-03 10:38:23   
457305      466422    8979081 2016-05-03 12:10:47   
457317      466434    8979405 2016-05-03 12:23:37   
457353      466470    8980137 2016-05-03 12:50:12   
457355      466472    8980162 2016-05-03 12:52:09   
457434      466551    8981857 2016-05-03 13:54:57   
457443      466560    8981933 2016-05-03 13:58:44   
457489      466606    8982910 2016-05-03 14:36:33   
457528      466645    8983623 2016-05-03 15:08:53   
457541      466658    8983853 2016-05-03 15:12:06   
470336      479453    9270382 2016-05-18 14:05:48   
472602      481719    9320260 2016-05-20 18:02:08   
475947      485064    9396390 2016-05-25 09:46:31   
475951      485068    9396398 2016-05-25 09:45:52   
475952      485069    9396399 2016-05-25 09:45:52   
475953      485070    9396404 2016-05-25 09:46:01   
475959      485076    9396500 2016-05-25 09:49:43   
475960      485077    9396501 2016-05-25 09:49:43   
475961      485078    9396502 2016-05-25 09:49:43   
478982      488099    9471204 2016-05-29 07:28:46   
478987      488104    9471209 2016-05-29 07:28:08   
478988      488105    9471210 2016-05-29 07:28:08   
478989      488106    9471212 2016-05-29 07:28:16   
478991      488108    9471250 2016-05-29 07:33:53   
478992      488109    9471251 2016-05-29 07:33:53   
478993      488110    9471252 2016-05-29 07:33:53   
485170      494287    9605167 2016-06-06 07:31:59   
485172      494289    9605169 2016-06-06 07:31:59   
485189      494306    9605449 2016-06-06 07:47:26   
485197      494314    9605498 2016-06-06 07:50:35   
524673      536550   10499304 2016-07-21 09:11:48   
524680      536557   10499374 2016-07-21 09:15:20   
524682      536559   10499377 2016-07-21 09:15:21   
524696      536573   10499556 2016-07-21 09:18:36   
530547      542424   10630152 2016-07-27 19:45:40   
530548      542425   10630161 2016

##### Engine Serial Number

In [15]:
faults_diagnostic['ecu_serial_number'].value_counts()

unknown            298549
6U13D13             11207
79845785            10302
79856768             8158
79845329             7199
79623056             7066
79621048             6828
79845786             6349
79840984             6345
79844876             5872
79623054             5663
79623410             5578
79844877             5506
79620769             5482
79844882             5417
79614871             5326
79615187             5218
79857688             5059
79857689             5019
79614865             4374
79619434             4206
79620768             4133
79615184             3857
79845331             3819
79619117             3797
79614866             3788
S381222841           3176
79615183             3169
79845327             3149
79844880             3113
79857687             3089
79857685             3065
00000000             3062
79620774             3055
79618850             3041
79619125             2855
79607068             2798
79620764             2798
79623055    

There are 'unknown' (298549), 'Unspecified' (97), and NoSerial (3) values. Let's combine them all into 'unknown'

In [16]:
faults_diagnostic['ecu_serial_number'] = faults_diagnostic['ecu_serial_number'].replace('Unspecified', 'unknown')
faults_diagnostic['ecu_serial_number'] = faults_diagnostic['ecu_serial_number'].replace('NoSerial', 'unknown')

##### Engine Model

In [17]:
faults_diagnostic['ecu_model'].sort_values().unique()

array(['0USA10_13_0405_2237A', '0USA13_13_0415_2238A', '202.35.0',
       '20412511P07', '6L u13D0890000000', '6U13D13', '6X1u10D1500000000',
       '6X1u13D1500000000', '6X1u17D1500000000', '6X1u20D1500000000',
       'CE', 'CECU3-NAMUX3', 'CECU3B-NAMUX4', 'E0031', 'EC60-adv',
       'EC80ESP', 'EC80ESP AM000036', 'EC80ESP AM000038', 'EC80ESP+',
       'EEO-xxF112C', 'FAOM-xx810S-EC3', 'Gen 4 Boot Loader', 'MX',
       'MX16U13D13', 'MX16U15D13', 'Y044053', 'Y049568',
       '________Y043718', 'unknown', nan], dtype=object)

Nothing to update for models

##### Engine Make

In [18]:
faults_diagnostic['ecu_make'].value_counts()

CMMNS              433403
unknown            298549
PACCR              277021
BNDWS               71001
PCAR                20229
EATON               12612
VOLVO                7252
?????                 755
????S                 627
????R                 589
?MMNS                 289
?CAR                  152
?ACCR                  39
???CR                  20
?NDWS                  15
?????MX16U13D13         9
?????MX                 6
??MNS                   3
?ATON                   3
???R                    1
??DWS                   1
5516014                 1
Name: ecu_make, dtype: int64

In [19]:
faults_diagnostic['ecu_make'] = faults_diagnostic['ecu_make'].replace('?????', 'unknown')
faults_diagnostic['ecu_make'] = faults_diagnostic['ecu_make'].replace('????S', 'unknown')
faults_diagnostic['ecu_make'] = faults_diagnostic['ecu_make'].replace('????R', 'PACCR')
faults_diagnostic['ecu_make'] = faults_diagnostic['ecu_make'].replace('?MMNS', 'CMMNS')
faults_diagnostic['ecu_make'] = faults_diagnostic['ecu_make'].replace('?CAR', 'PCAR')
faults_diagnostic['ecu_make'] = faults_diagnostic['ecu_make'].replace('?ACCR', 'PACCR')
faults_diagnostic['ecu_make'] = faults_diagnostic['ecu_make'].replace('???CR', 'PACCR')
faults_diagnostic['ecu_make'] = faults_diagnostic['ecu_make'].replace('?NDWS', 'BNDWS')
faults_diagnostic['ecu_make'] = faults_diagnostic['ecu_make'].replace('?????MX16U13D13', 'MX16U13D13')
faults_diagnostic['ecu_make'] = faults_diagnostic['ecu_make'].replace('?????MX', 'MX')
faults_diagnostic['ecu_make'] = faults_diagnostic['ecu_make'].replace('?ATON', 'EATON')
faults_diagnostic['ecu_make'] = faults_diagnostic['ecu_make'].replace('??MNS', 'CMMNS')
faults_diagnostic['ecu_make'] = faults_diagnostic['ecu_make'].replace('???R', 'PCAR')
faults_diagnostic['ecu_make'] = faults_diagnostic['ecu_make'].replace('??DWS', 'BNDWS')

##### ECU Source

In [20]:
faults_diagnostic['ecu_source'].value_counts()

0     528044
49    514059
11    131122
3      13484
61       626
Name: ecu_source, dtype: int64

Nothing to clean

##### SPN

In [21]:
faults_diagnostic['spn'].value_counts()

111       365489
929       256541
96         90041
829        87788
639        41062
97         26745
596        22571
50353      11773
1569       10927
1761        9981
2863        8233
789         8022
629         8012
1068        7994
791         7799
1231        6992
1067        6947
641         6292
91          6210
3226        5547
3216        5420
37          5183
792         5164
412         5044
790         5011
807         4945
886         4190
1807        4081
627         4043
802         4036
171         4036
611         4026
1059        3952
51923       3950
793         3906
3251        3905
3464        3382
4364        2618
102         2578
630         2460
157         2426
70          2268
5396        2223
1483        2178
5848        2100
3610        2021
1045        1994
907         1962
4096        1877
5444        1800
1209        1736
523531      1718
2623        1700
101         1686
110         1652
3031        1650
1808        1636
934         1613
248         16

Nothing to clean

##### FMI

In [22]:
faults_diagnostic['fmi'].value_counts()

17    326553
9     288893
3     188631
2      82334
18     53602
31     39881
4      39734
0      27577
7      23249
1      20887
15     20739
14     12699
12     12233
5       9683
16      9242
10      7437
8       5741
19      5733
20      4440
11      3847
13      2634
6       1027
23       396
21       131
29         8
22         4
Name: fmi, dtype: int64

Nothing to clean

##### Active

In [23]:
faults_diagnostic['active'].value_counts(normalize=True)

True     0.512454
False    0.487546
Name: active, dtype: float64

Nothing to clean

##### Active Count

In [24]:
faults_diagnostic['active_transition_count'].value_counts()

126    581874
1      305476
127    102512
2       45075
3       13019
4        9658
6        6880
5        6777
0        5781
7        4782
8        4088
9        3750
11       3387
10       3326
12       2713
13       2603
14       2420
15       2358
16       2158
17       2064
18       1914
20       1755
19       1753
21       1590
22       1482
23       1475
24       1459
25       1393
26       1391
28       1277
27       1276
29       1230
30       1164
33       1132
32       1128
31       1074
34       1039
36       1000
39        973
40        972
37        970
35        968
38        965
44        915
43        908
42        906
41        896
45        837
48        832
46        813
47        811
50        800
49        798
55        762
52        747
54        738
51        731
53        728
59        723
58        720
56        689
57        687
63        676
61        671
64        655
70        648
62        644
60        624
66        623
72        618
67        618
68    

Nothing to clean

##### Equipment ID

In [25]:
faults_diagnostic['equipment_id'].value_counts()

1641          17492
1605          16393
1646          15462
1618          14986
1606          14973
1619          14832
1625          14783
1645          14268
1644          13920
1649          13557
1610          13489
1630          13149
1647          13117
1634          13003
1609          12902
1642          12653
1623          11834
1611          11804
1814          10361
1692          10301
1612           9554
1816           8164
1809           8026
1939           7966
1622           7918
1806           7771
1601           7437
1603           7141
1469           7116
1592           7085
1749           7000
1808           6856
1815           6474
1803           6390
1437           6133
309            5991
305            5738
1600           5684
1804           5670
1589           5532
1562           5483
1620           5470
1995           5365
1559           5358
1818           5135
1820           5080
1616           5015
1810           4883
1422           4736
1873           4584


Per the instructions, we'll need to remove any equipment_id that is greater than 5 digits

In [26]:
faults_diagnostic = faults_diagnostic[faults_diagnostic['equipment_id'].str.len() <= 5]

This removes ~2k rows

##### MCT Number

In [27]:
faults_diagnostic['mct_number'].value_counts()

105415080    16503
105420184    15507
105381514    15330
105415457    15119
105416377    15034
105417985    13590
105442858    13187
105420195    13141
105415391    13079
105338710    11834
105420590    11038
105357098    10719
105369518    10427
105420520     9939
105320363     9580
105430420     8590
105420005     8532
105465351     8361
105420580     8111
105434215     8086
105351093     8064
105330008     8019
105304730     8018
105438415     8005
105415630     7889
105364937     7479
105357794     7265
105411041     7187
105357909     6937
105400037     6929
105438416     6877
105355660     6723
105333857     6639
105460307     6529
105432778     6421
105356370     6406
105442799     6200
105435663     6040
105334108     5978
105392933     5872
105410475     5784
105356577     5732
105338555     5726
105406655     5722
105362919     5675
105460271     5670
105410840     5610
105412415     5316
105349500     5164
105465571     5080
105465373     5054
105411109     5009
105427242   

Nothing to clean

##### Latitude and Longitude

In [28]:
faults_diagnostic[faults_diagnostic['latitude'] == 0]

record_id     ess_id     event_timestamp  \
1150008    1207232  106294736 2019-06-24 09:10:10   
1150009    1207233  106294737 2019-06-24 09:10:09   
1150010    1207234  106294738 2019-06-24 09:10:11   
1150011    1207235  106294739 2019-06-24 09:10:12   
1172275    1231439  116418325 2020-01-09 20:50:32   
1172276    1231440  116418326 2020-01-09 20:50:32   

                                         event_description  \
1150008              Low Voltage (Sensor supply voltage 2)   
1150009  Low Voltage (Engine Exhaust Gas Recirculation ...   
1150010                                                NaN   
1150011         Low Current Aftertreatment Fuel Injector 1   
1172275                          High Voltage (Fuel Level)   
1172276              High Voltage (Left Fuel Level Sensor)   

        ecu_software_version ecu_serial_number      ecu_model ecu_make  \
1150008              unknown           unknown        unknown  unknown   
1150009              unknown           unknown        unknown  unknown   
1150010                  NaN               NaN            NaN      NaN   
1150011              unknown           unknown        unknown  unknown   
1172275                  NaN               NaN  CECU3B-NAMUX4    PACCR   
1172276                  NaN               NaN  CECU3B-NAMUX4    PACCR   

         ecu_source   spn  fmi  active  active_transition_count equipment_id  \
1150008           0  3510    4    True                        1         2007   
1150009           0    27    4    True                        1         2007   
1150010          11  1807   12    True                        1         2007   
1150011           0  3556    5    True                        1         2007   
1172275          49    96    3    True                      126         1857   
1172276          49   829    3    True                      126         1857   

         mct_number  latitude  longitude  location_timestamp  fault_id  \
1150008   105304730       0.0        0.0 2019-06-24 08:43:30   1207232   
1150009   105304730       0.0        0.0 2019-06-24 08:43:30   1207233   
1150010   105304730       0.0        0.0 2019-06-24 08:43:30   1207234   
1150011   105304730       0.0        0.0 2019-06-24 08:43:30   1207235   
1172275   105430420       0.0        0.0 2020-01-09 20:51:09   1231439   
1172276   105430420       0.0        0.0 2020-01-09 20:51:09   1231440   

        accelerator_pedal barometric_pressure cruise_control_active  \
1150008                 0               14.21                 False   
1150009                 0               14.21                 False   
1150010                 0               14.21                 False   
1150011              25.2               14.21                 False   
1172275                 0                14.5                 False   
1172276                 0                14.5                 False   

        cruise_control_set_speed distance_ltd engine_coolant_temperature  \
1150008                  64.6226       288318                        203   
1150009                  64.6226       288318                        203   
1150010                  64.6226       288318                        203   
1150011                  64.6226       288318                        203   
1172275                 66.48672     383791.8                      102.2   
1172276                 66.48672     383791.8                      102.2   

        engine_load engine_oil_pressure engine_oil_temperature engine_rpm  \
1150008           0                   0                 216.05          0   
1150009           0                   0                 216.05          0   
1150010           0                   0               215.9937          0   
1150011           0                   0               215.9937          0   
1172275          10                40.6               102.5375    600.125   
1172276          10                40.6               102.5375    600.125   

        engine_time_ltd fuel_lev

8 rows contain a latitude of 0.0

In [29]:
faults_diagnostic[faults_diagnostic['longitude'] == 0]

record_id     ess_id     event_timestamp  \
1150008    1207232  106294736 2019-06-24 09:10:10   
1150009    1207233  106294737 2019-06-24 09:10:09   
1150010    1207234  106294738 2019-06-24 09:10:11   
1150011    1207235  106294739 2019-06-24 09:10:12   
1172275    1231439  116418325 2020-01-09 20:50:32   
1172276    1231440  116418326 2020-01-09 20:50:32   

                                         event_description  \
1150008              Low Voltage (Sensor supply voltage 2)   
1150009  Low Voltage (Engine Exhaust Gas Recirculation ...   
1150010                                                NaN   
1150011         Low Current Aftertreatment Fuel Injector 1   
1172275                          High Voltage (Fuel Level)   
1172276              High Voltage (Left Fuel Level Sensor)   

        ecu_software_version ecu_serial_number      ecu_model ecu_make  \
1150008              unknown           unknown        unknown  unknown   
1150009              unknown           unknown        unknown  unknown   
1150010                  NaN               NaN            NaN      NaN   
1150011              unknown           unknown        unknown  unknown   
1172275                  NaN               NaN  CECU3B-NAMUX4    PACCR   
1172276                  NaN               NaN  CECU3B-NAMUX4    PACCR   

         ecu_source   spn  fmi  active  active_transition_count equipment_id  \
1150008           0  3510    4    True                        1         2007   
1150009           0    27    4    True                        1         2007   
1150010          11  1807   12    True                        1         2007   
1150011           0  3556    5    True                        1         2007   
1172275          49    96    3    True                      126         1857   
1172276          49   829    3    True                      126         1857   

         mct_number  latitude  longitude  location_timestamp  fault_id  \
1150008   105304730       0.0        0.0 2019-06-24 08:43:30   1207232   
1150009   105304730       0.0        0.0 2019-06-24 08:43:30   1207233   
1150010   105304730       0.0        0.0 2019-06-24 08:43:30   1207234   
1150011   105304730       0.0        0.0 2019-06-24 08:43:30   1207235   
1172275   105430420       0.0        0.0 2020-01-09 20:51:09   1231439   
1172276   105430420       0.0        0.0 2020-01-09 20:51:09   1231440   

        accelerator_pedal barometric_pressure cruise_control_active  \
1150008                 0               14.21                 False   
1150009                 0               14.21                 False   
1150010                 0               14.21                 False   
1150011              25.2               14.21                 False   
1172275                 0                14.5                 False   
1172276                 0                14.5                 False   

        cruise_control_set_speed distance_ltd engine_coolant_temperature  \
1150008                  64.6226       288318                        203   
1150009                  64.6226       288318                        203   
1150010                  64.6226       288318                        203   
1150011                  64.6226       288318                        203   
1172275                 66.48672     383791.8                      102.2   
1172276                 66.48672     383791.8                      102.2   

        engine_load engine_oil_pressure engine_oil_temperature engine_rpm  \
1150008           0                   0                 216.05          0   
1150009           0                   0                 216.05          0   
1150010           0                   0               215.9937          0   
1150011           0                   0               215.9937          0   
1172275          10                40.6               102.5375    600.125   
1172276          10                40.6               102.5375    600.125   

        engine_time_ltd fuel_lev

The same 8 rows contain a longitude of 0.0

We'll replace 0.0 with np.nan. We'll get some values from ffill and backfills before filtering out service stations by latitude and longitude.

In [30]:
lat_long = ['latitude', 'longitude']

for column in lat_long:
    faults_diagnostic[column] = faults_diagnostic[column].replace('0.0', np.nan)

##### Location Timestamp

In [31]:
faults_diagnostic['location_timestamp'].dt.year.value_counts()

2016    332288
2015    325678
2017    254860
2018    143423
2019    111520
2020     17196
2000       192
2011         8
2014         1
Name: location_timestamp, dtype: int64

We'll need to investigate the records pre-2014

Is there a way to compare location_timestamp to event_timestamp? More importantly, is this worth doing?

In [32]:
faults_diagnostic[faults_diagnostic['location_timestamp'].dt.year == 2000]

record_id     ess_id     event_timestamp  \
1154193    1211417  108604425 2000-03-18 19:14:10   
1154194    1211418  108604426 2000-03-18 19:14:10   
1154195    1211419  108604487 2000-03-18 19:20:47   
1154196    1211420  108604488 2000-03-18 19:20:47   
1154198    1211422  108608408 2000-03-19 02:59:58   
1154199    1211423  108608954 2000-03-19 03:58:23   
1154208    1211432  108615304 2000-03-19 07:32:53   
1154209    1211433  108615321 2000-03-19 07:34:31   
1154211    1211435  108620729 2000-03-19 08:40:03   
1154212    1211436  108620730 2000-03-19 08:40:03   
1154215    1211439  108622089 2000-03-19 08:58:27   
1154216    1211440  108622090 2000-03-19 08:58:27   
1154219    1211443  108622593 2000-03-19 09:03:22   
1154220    1211444  108622594 2000-03-19 09:03:22   
1154223    1211447  108630527 2000-03-19 10:35:14   
1154224    1211448  108630528 2000-03-19 10:35:14   
1154225    1211449  108631659 2000-03-19 10:51:28   
1154228    1211452  108632613 2000-03-19 11:11:02   
1154243    1211467  108634615 2000-03-19 11:42:06   
1154244    1211468  108634616 2000-03-19 11:42:06   
1154245    1211469  108635168 2000-03-19 11:50:17   
1154246    1211470  108635169 2000-03-19 11:50:17   
1154247    1211471  108635373 2000-03-19 11:55:06   
1154248    1211472  108635374 2000-03-19 11:55:06   
1154249    1211473  108636257 2000-03-19 12:08:22   
1154251    1211475  108638110 2000-03-19 12:34:51   
1154252    1211476  108638111 2000-03-19 12:34:51   
1154253    1211477  108638380 2000-03-19 12:38:20   
1154254    1211478  108638381 2000-03-19 12:38:20   
1154255    1211479  108639463 2000-03-19 12:55:01   
1154256    1211480  108639464 2000-03-19 12:55:01   
1154257    1211481  108639840 2000-03-19 13:02:36   
1154258    1211482  108640040 2000-03-19 13:06:49   
1154259    1211483  108640041 2000-03-19 13:06:49   
1154260    1211484  108642452 2000-03-19 13:41:24   
1154261    1211485  108642453 2000-03-19 13:41:24   
1154263    1211487  108644439 2000-03-19 14:02:34   
1154264    1211488  108645372 2000-03-19 14:07:55   
1154265    1211489  108645850 2000-03-19 14:11:59   
1154266    1211490  108646591 2000-03-19 14:17:26   
1154271    1211495  108651978 2000-03-19 15:12:24   
1154272    1211496  108651979 2000-03-19 15:12:24   
1154275    1211499  108653960 2000-03-19 15:35:59   
1154276    1211500  108653961 2000-03-19 15:35:59   
1154277    1211501  108654236 2000-03-19 15:37:49   
1154278    1211502  108654338 2000-03-19 15:39:46   
1154279    1211503  108654582 2000-03-19 15:42:29   
1154280    1211504  108654583 2000-03-19 15:42:30   
1154281    1211505  108655772 2000-03-19 16:05:58   
1154282    1211506  108655773 2000-03-19 16:05:58   
1154283    1211507  108655839 2000-03-19 16:06:32   
1154284    1211508  108656295 2000-03-19 16:13:43   
1154285    1211509  108656621 2000-03-19 16:20:31   
1154286    1211510  108656622 2000-03-19 16:20:31   
1154287    1211511  108656781 2000-03-19 16:25:25   
1154288    1211512  108656844 2000-03-19 16:27:13   
1154293    1211517  108659676 2000-03-19 17:06:22   
1154296    1211520  108659778 2000-03-19 17:08:54   
1154297    1211521  108661657 2000-03-19 17:28:04   
1154298    1211522  108661658 2000-03-19 17:27:44   
1154299    1211523  108661659 2000-03-19 17:27:44   
1154300    1211524  108661660 2000-03-19 17:27:58   
1154301    1211525  108661886 2000-03-19 17:55:53   
1154302    1211526  108662038 2000-03-19 17:59:01   
1154303    1211527  108662262 2000-03-19 18:09:55   
1154304    1211528  108662312 2000-03-19 18:12:16   
1154305    1211529  108662313 2000-03-19 18:12:16   
1154306    1211530  108662645 2000-03-19 18:02:53   
1154307    1211531  108662646 2000-03-19 18:02:53   
1154308    1211532  108662647 2000-03-19 18:02:53   
1154309    1211533  108663626 2000-03-19 18:56:09   
1154310    1211534  108663627 2000-03-19 18:56:09   
1154311    1211535  108664021 2000-03-19 19:13:54   
1154312    1211536  108664067 2000-03-19 19:17:46   
1154313    1211537  108665697 2000

In [33]:
faults_diagnostic[faults_diagnostic['location_timestamp'].dt.year == 2000]['equipment_id'].unique()

array(['2015', '1849', '2283', '2034', '1909', '1997', '1961', '1826',
       '1887', '1968', '1873', '2074', '1965', '2336', '1977', '1970',
       '1869', '1861', '1979', '2037', '2184', '2054', '1989', '2159',
       '2158', '2174'], dtype=object)

In [34]:
faults_diagnostic[faults_diagnostic['equipment_id'] == '2015']

record_id     ess_id     event_timestamp  \
768584      787340   16361053 2017-05-18 05:18:34   
768589      787345   16361224 2017-05-18 05:32:49   
780578      800135   17117698 2017-06-03 16:57:41   
780595      800152   17119828 2017-06-03 17:57:41   
780601      800158   17120589 2017-06-03 18:24:23   
780619      800176   17122776 2017-06-03 19:41:11   
1000018    1037155   55341060 2018-07-17 05:25:26   
1000020    1037157   55341160 2018-07-17 05:32:26   
1009948    1047085   58819006 2018-08-18 09:47:45   
1010006    1047143   58837028 2018-08-18 13:35:46   
1010072    1047209   58865909 2018-08-19 01:09:45   
1010082    1047219   58870103 2018-08-19 05:49:10   
1010333    1048427   58967791 2018-08-20 09:49:32   
1010580    1048674   59052643 2018-08-21 02:19:01   
1010638    1048732   59075852 2018-08-21 07:52:13   
1010961    1049055   59184980 2018-08-22 01:54:58   
1010994    1049088   59193637 2018-08-22 05:20:45   
1016458    1054552   61218796 2018-09-10 13:16:51   
1016535    1054629   61250851 2018-09-10 17:13:16   
1017067    1055161   61486364 2018-09-12 11:33:46   
1017107    1055201   61504766 2018-09-12 13:34:36   
1017905    1055999   61779993 2018-09-15 02:48:35   
1017975    1056069   61802866 2018-09-15 10:28:05   
1018368    1056462   61924343 2018-09-17 08:23:55   
1018425    1056519   61940904 2018-09-17 10:29:33   
1019230    1058225   62271045 2018-09-20 00:03:56   
1019301    1058296   62286713 2018-09-20 06:16:47   
1020158    1059153   62576840 2018-09-23 03:28:51   
1027668    1066663   65059687 2018-10-15 15:40:29   
1027777    1066772   65112741 2018-10-16 07:42:31   
1027781    1066776   65114789 2018-10-16 07:56:46   
1027782    1066777   65114790 2018-10-16 07:56:45   
1027784    1066779   65115584 2018-10-16 08:04:31   
1027785    1066780   65115587 2018-10-16 08:04:39   
1027797    1066792   65119091 2018-10-16 08:29:55   
1065622    1109375   74752049 2019-01-20 06:37:13   
1065739    1109492   74774250 2019-01-20 13:11:24   
1068107    1111860   75394247 2019-01-26 07:14:57   
1068108    1111861   75394248 2019-01-26 07:16:51   
1068591    1112344   75523186 2019-01-28 06:45:22   
1068662    1112415   75544740 2019-01-28 09:48:26   
1070116    1113869   75893659 2019-01-31 07:20:24   
1070117    1113870   75893769 2019-01-31 07:22:36   
1117267    1166763   89452625 2019-06-05 17:59:04   
1117268    1166764   89452626 2019-06-05 17:59:04   
1117269    1166765   89452627 2019-06-05 17:59:05   
1117270    1166766   89452628 2019-06-05 17:59:04   
1120261    1170683   90692108 2019-06-15 17:50:30   
1120262    1170684   90692245 2019-06-15 17:50:51   
1120263    1170685   90692246 2019-06-15 17:51:35   
1120264    1170686   90692247 2019-06-15 17:51:35   
1120265    1170687   90692248 2019-06-15 17:51:35   
1120266    1170688   90692249 2019-06-15 17:50:55   
1120267    1170689   90692420 2019-06-15 17:53:22   
1120268    1170690   90692421 2019-06-15 17:52:42   
1120269    1170691   90692422 2019-06-15 17:53:28   
1120274    1170696   90692772 2019-06-15 17:59:28   
1120277    1170699   90692897 2019-06-15 18:04:54   
1149399    1205627  105860515 2019-10-12 22:40:10   
1149400    1205628  105860516 2019-10-12 22:39:31   
1150006    1207230  106294106 2019-10-16 07:25:55   
1150007    1207231  106294107 2019-10-16 07:25:55   
1150181    1207405  106324924 2019-10-16 10:27:51   
1150182    1207406  106324925 2019-10-16 10:27:51   
1152872    1210096  107976271 2019-10-29 06:24:15   
1152873    1210097  107976272 2019-10-29 06:24:15   
1152894    1210118  107983817 2019-10-29 07:29:00   
1152895    1210119  107983818 2019-10-29 07:29:00   
1152966    1210190  108021037 2019-10-29 10:28:18   
1152967    1210191  108021038 2019-10-29 10:28:18   
1152968    1210192  108021863 2019-10-29 10:32:37   
1152969    1210193  108021864 2019-10-29 10:32:37   
1153000    1210224  108029509 2019-10-29 10:40:30   
1153001    1210225  108029510 2019-10-29 10:40:30   
1153015    1210239  108039739 2019

Interesting...the 2000 records appear to have the event_timestamps and location_timestamps in the same year

In [35]:
faults_diagnostic[faults_diagnostic['location_timestamp'].dt.year == 2011]

record_id     ess_id     event_timestamp  \
1154812    1212036  108848937 2011-01-02 03:17:37   
1155217    1212441  109079376 2011-01-01 05:10:23   
1155218    1212442  109079377 2011-01-01 05:09:43   
1155219    1212443  109079378 2011-01-01 05:11:26   
1155230    1212454  109079376 2011-01-01 05:10:23   
1155231    1212455  109079377 2011-01-01 05:09:43   
1155232    1212456  109079378 2011-01-01 05:11:26   
1171320    1230484  115919584 2011-01-01 00:10:32   

                                         event_description  \
1154812             Low (Severity Low) Catalyst Tank Level   
1155217  Special Instructions Transmission Clutch Actuator   
1155218  Low (Severity Medium) Transmission Air Tank Pr...   
1155219  Low (Severity Medium) Transmission Air Tank Pr...   
1155230  Special Instructions Transmission Clutch Actuator   
1155231  Low (Severity Medium) Transmission Air Tank Pr...   
1155232  Low (Severity Medium) Transmission Air Tank Pr...   
1171320  Low (Severity High) Transmission Air Tank Pres...   

                                      ecu_software_version ecu_serial_number  \
1154812    04384413*22059242*090617144354*60701715*G1*BGT*          80015667   
1155217  5516018*202.56.0*5516502*E003.e003*5539540*25....          Z0018064   
1155218  5516018*202.56.0*5516502*E003.e003*5539540*25....          Z0018064   
1155219  5516018*202.56.0*5516502*E003.e003*5539540*25....          Z0018064   
1155230  5516018*202.56.0*5516502*E003.e003*5539540*25....          Z0018064   
1155231  5516018*202.56.0*5516502*E003.e003*5539540*25....          Z0018064   
1155232  5516018*202.56.0*5516502*E003.e003*5539540*25....          Z0018064   
1171320  5516014*202.35.0*5516502*E003.e003*5539478*25....          Z0047881   

                 ecu_model ecu_make  ecu_source   spn  fmi  active  \
1154812  6X1u17D1500000000    CMMNS           0  1761   17    True   
1155217        EEO-xxF112C    EATON           3   788   14   False   
1155218        EEO-xxF112C    EATON           3    37   18    True   
1155219        EEO-xxF112C    EATON           3    37   18   False   
1155230        EEO-xxF112C    EATON           3   788   14   False   
1155231        EEO-xxF112C    EATON           3    37   18    True   
1155232        EEO-xxF112C    EATON           3    37   18   False   
1171320        EEO-xxF112C    EATON           3    37    1    True   

         active_transition_count equipment_id  mct_number   latitude  \
1154812                        5         2144   105418094  35.153101   
1155217                        2         2218   105435718  40.226851   
1155218                        5         2218   105435718  40.226851   
1155219                        5         2218   105435718  40.226851   
1155230                        2         2218   105435718  40.226851   
1155231                        5         2218   105435718  40.226851   
1155232                        5         2218   105435718  40.226851   
1171320                        3         2283   105437713  28.156342   

         longitude  location_timestamp  fault_id accelerator_pedal  \
1154812 -86.592037 2011-01-02 03:18:13   1212036                 0   
1155217 -76.721296 2011-01-01 05:10:19   1212441               NaN   
1155218 -76.721296 2011-01-01 05:10:20   1212442                 0   
1155219 -76.721296 2011-01-01 05:11:22   1212443               NaN   
1155230 -76.721296 2011-01-01 05:10:19   1212454               NaN   
1155231 -76.721296 2011-01-01 05:10:20   1212455                 0   
1155232 -76.721296 2011-01-01 05:11:22   1212456               NaN   
1171320 -81.799861 2011-01-01 00:11:08   1230484                 0   

        barometric_pressure cruise_control_active cruise_control_set_speed  \
1154812             14.4275                 False                        0   
1155217                 NaN                   NaN                      NaN   
1155218                14.5                 False                        0   
1155219                 NaN 

In [36]:
faults_diagnostic[faults_diagnostic['location_timestamp'].dt.year == 2014]

record_id   ess_id     event_timestamp  \
27724      28607  1994354 2015-04-21 03:31:18   

                                       event_description ecu_software_version  \
27724  Abnormal Rate of Change Aftertreatment 1 Outle...              unknown   

      ecu_serial_number ecu_model ecu_make  ecu_source   spn  fmi  active  \
27724           unknown   unknown  unknown           0  3226   10    True   

       active_transition_count equipment_id  mct_number   latitude  longitude  \
27724                        1         1470   105446251  36.945092 -80.944583   

       location_timestamp  fault_id accelerator_pedal barometric_pressure  \
27724 2014-12-30 13:44:03     28607               NaN                 NaN   

      cruise_control_active cruise_control_set_speed distance_ltd  \
27724                   NaN                      NaN          NaN   

      engine_coolant_temperature engine_load engine_oil_pressure  \
27724                        NaN         NaN                 NaN   

      engine_oil_temperature engine_rpm engine_time_ltd fuel_level fuel_ltd  \
27724                    NaN        NaN             NaN        NaN      NaN   

      fuel_rate fuel_temperature ign_status intake_manifold_temperature  \
27724       NaN              NaN       True                         NaN   

      lamp_status parking_brake service_distance speed  \
27724       17407           NaN              NaN   NaN   

      switched_battery_voltage throttle turbo_boost_pressure  
27724                      NaN      NaN                  NaN

Given the discrepancies, it may make sense to use only one timestamp column

##### Fault ID

'fault_id' was used to join the datasets and is redundant with 'record_id'. We'll drop it now

In [37]:
faults_diagnostic = faults_diagnostic.drop(columns=['fault_id'])

##### Accelerator Pedal

In [38]:
faults_diagnostic['accelerator_pedal'].value_counts()

0           276459
100          53985
99.6          2841
0.4           1939
27.2          1910
27.6          1865
28.4          1862
28            1823
28.8          1815
25.2          1797
26.8          1796
26.4          1792
29.6          1788
25.6          1771
26            1756
24.4          1722
30.8          1716
29.2          1715
31.6          1686
30.4          1677
32            1666
32.8          1649
22.8          1639
31.2          1631
23.6          1618
33.2          1615
33.6          1601
23.2          1594
24            1588
32.4          1578
30            1578
24.8          1567
34            1517
22            1503
21.6          1502
0.8           1497
34.4          1493
22.4          1475
35.2          1464
21.2          1419
20.8          1392
20.4          1369
36.4          1350
36.8          1338
99.2          1326
35.6          1325
36            1310
34.8          1282
37.6          1281
37.2          1270
19.6          1251
38.4          1249
19.2        

Several values have commas—this is consistent across the diagnostic data, so we'll clean those all up at once

##### Barometric Pressure

In [39]:
faults_diagnostic['barometric_pressure'].value_counts()

14.355      85702
14.4275     82051
14.2825     76797
14.5        56810
14.21       56652
14.5725     40160
14.1375     37609
14.645      32297
14.065      21652
14.7175     20379
13.9925     11503
14.79        7775
13.92        7209
13.775       6653
13.8475      6545
13.7025      6523
13.63        5847
13.5575      4342
3.5525       3325
13.485       3113
14.8625      2420
13.4125      2111
3.625        1448
13.34        1129
3.48          822
14.935        685
13.2675       598
12.2525       311
3.4075        280
13.195        237
15.0075       182
7.25          182
13.1225       154
3.6975        121
3.335         118
13.05         109
12.9775       104
12.615         87
12.8325        70
12.6875        60
12.905         51
11.6725        43
12.76          42
12.5425        40
12.325         39
12.3975        34
11.9625        33
12.18          33
11.455         31
11.6           31
12.035         30
11.745         27
11.5275        26
12.47          26
11.89          24
15.08     

Commas to be fixed

##### Cruise Control Active

In [40]:
faults_diagnostic['cruise_control_active'].value_counts(normalize=True)

False    0.911834
True     0.088166
Name: cruise_control_active, dtype: float64

##### Cruise Control Set Speed

In [41]:
faults_diagnostic['cruise_control_set_speed'].value_counts()

66.48672    278090
64.6226     130383
0            34280
65.24397     10952
65.86535      9675
64.00124      9302
59.65163      8065
63.37986      6705
61.51575      6639
60.89438      6628
60.27301      5857
55.30204      5697
59.03026      5294
62.75849      5226
56.54478      4805
57.78752      4575
57.16615      4319
62.13712      4262
54.68066      4237
55.92341      4223
58.40889      4218
67.10809      4060
42.25324      2270
54.05929      2212
53.43792       935
68.35083       902
52.81655       747
50.33107       742
68.97221       669
51.57381       626
49.70969       537
52.19518       508
45.3601        465
44.73873       464
47.84558       451
45.98147       440
46.60284       410
41.63187       410
44.11736       387
50.95244       373
47.22421       345
67.72946       341
49.08833       304
70.83632       275
48.46695       242
27.9617        199
70.21494       198
31.68993       191
40.38913       190
43.49598       160
29.82582       150
29.20445       139
34.79679    

There are 34,698 records that have cruise control set to 0 mph. Interesting

##### Distance (Lifetime To Date)

In [42]:
faults_diagnostic['distance_ltd']

0             423178.7
1                  NaN
2                  NaN
3                  NaN
4                  NaN
              ...     
1187330            NaN
1187331       423937.9
1187332       465925.4
1187333    28606.65625
1187334            NaN
Name: distance_ltd, Length: 1185166, dtype: object

We'll need to fix commas before going further

##### Engine Coolant Temperature

In [43]:
faults_diagnostic['engine_coolant_temperature'].value_counts()

185      69413
183.2    60148
186.8    56234
181.4    40461
188.6    36754
179.6    25905
190.4    24036
177.8    16864
192.2    14203
174.2    12874
176      12197
172.4    10172
170.6     8488
194       7097
168.8     6848
167       5953
195.8     5240
165.2     4803
163.4     4519
197.6     4374
161.6     4032
199.4     3855
159.8     3469
156.2     3460
158       3225
154.4     3157
201.2     3052
150.8     2851
152.6     2824
86        2797
87.8      2773
203       2736
91.4      2706
95        2646
98.6      2624
96.8      2605
93.2      2588
84.2      2587
100.4     2581
89.6      2536
102.2     2514
80.6      2481
149       2444
120.2     2439
78.8      2427
82.4      2394
147.2     2352
77        2344
105.8     2339
109.4     2329
204.8     2317
104       2299
138.2     2289
114.8     2288
107.6     2287
116.6     2275
75.2      2273
118.4     2212
132.8     2209
145.4     2184
111.2     2175
136.4     2169
113       2160
134.6     2145
127.4     2118
143.6     2118
73.4      

There are a few negative values and also more commas

They can go to -36 which does not occur in the dataset. 

##### Engine Load

In [44]:
faults_diagnostic['engine_load'].value_counts()

0      90770
100    34976
12     25335
14     24209
16     22158
10     18771
17     18567
13     17329
11     15876
15     15855
19     14054
21     11516
18     10894
9       9689
23      9321
8       8731
20      8206
25      7598
26      7283
28      7184
22      7046
1       6839
30      6660
7       5710
24      5589
32      5006
27      4920
29      4814
99      4590
33      4516
35      4131
31      3953
37      3776
39      3654
41      3564
42      3448
44      3354
46      3308
48      3260
51      3071
50      3005
6       2961
34      2930
53      2900
58      2733
36      2725
55      2719
5       2715
57      2714
60      2543
62      2542
66      2461
40      2456
38      2451
64      2381
69      2314
45      2284
43      2253
67      2238
47      2197
49      2116
98      2104
73      2064
52      2047
71      2033
75      2022
78      1999
76      1978
54      1914
56      1911
3       1877
80      1863
59      1783
82      1772
61      1752
63      1685
83      1647

Nothing to clean

##### Engine Oil Pressure

In [45]:
faults_diagnostic['engine_oil_pressure'].value_counts()

0           43115
38.86       29287
38.28       28559
39.44       26958
37.7        26588
37.12       24910
36.54       23697
40.02       22957
35.96       21914
35.38       20106
40.6        18270
34.8        16546
41.18       15029
34.22       13391
41.76       12058
33.64       10613
42.34        9820
20.88        9518
21.46        9394
20.3         9131
22.04        8901
19.72        8281
22.62        8055
42.92        7896
23.2         7622
33.06        7619
19.14        7007
23.78        6896
43.5         6299
24.36        6100
24.94        5358
32.48        5307
18.56        5241
44.08        5165
25.52        4826
26.1         4520
44.66        4253
31.9         4212
26.68        4176
27.26        3945
45.24        3746
17.98        3665
30.16        3617
31.32        3583
30.74        3511
27.84        3497
29.58        3426
29           3366
28.42        3270
45.82        3119
17.4         2927
46.4         2668
46.98        2276
16.82        2135
47.56        1900
48.14     

Commas to replace

##### Engine Oil Temperature

In [46]:
faults_diagnostic['engine_oil_temperature'].value_counts()

215.9937     4114
217.4563     4047
216.725      4022
213.7437     3854
214.5313     3834
218.2437     3833
219.7063     3813
210.5375     3679
215.2625     3637
213.0125     3604
211.6625     3603
218.975      3591
212.2813     3589
211.1        3467
220.4375     3438
221.1687     3220
210.0313     2982
221.9        2969
209.4688     2935
208.9063     2762
215.825      2758
217.2875     2753
211.55       2749
207.7813     2668
216.5562     2600
219.5375     2596
212.1125     2566
215.0375     2554
208.3438     2538
207.2188     2508
206.6563     2506
222.6875     2504
212.8438     2501
214.3062     2416
210.9875     2406
213.575      2373
206.0938     2371
220.2688     2352
205.5313     2349
210.425      2335
218.75       2328
218.0188     2293
204.9688     2278
221          2259
204.4063     2246
209.8625     2227
224.15       2220
203.8438     2175
223.4187     2125
209.3        2117
203.2813     2096
202.775      2030
206.5437     1937
224.8813     1929
197.6        1865
208.7375  

Have some negative values and commas

##### Engine RPM

In [47]:
faults_diagnostic['engine_rpm'].value_counts()

0          45563
600.125     5367
599.75      5192
600.375     5145
599.875     5136
           ...  
174            1
181            1
497.625        1
198.875        1
413.5          1
Name: engine_rpm, Length: 11470, dtype: int64

In [48]:
faults_diagnostic[faults_diagnostic['engine_rpm'].str.contains(",", regex=False, na=False)]

record_id    ess_id     event_timestamp  \
17519       18402   1823305 2015-04-10 14:31:59   
57330       59400   2517346 2015-05-19 20:48:51   
98748      101134   3180493 2015-06-27 05:46:01   
109858     112244   3326811 2015-07-07 05:31:43   
130411     132797   3623872 2015-07-22 22:17:35   
167370     169756   4093848 2015-08-17 09:36:38   
190360     192746   4441365 2015-09-04 04:46:23   
275289     279109   5795315 2015-11-16 03:04:21   
277270     281776   5827502 2015-11-17 08:23:45   
278386     282892   5845487 2015-11-18 07:58:38   
282142     286648   5909821 2015-11-20 22:30:20   
533850     545727  10702825 2016-08-01 02:09:42   

                                        event_description  \
17519   Data May Be Invalid Relative Speed; Rear Axle ...   
57330   Data May Be Invalid Relative Speed; Rear Axle ...   
98748   Data May Be Invalid Relative Speed; Rear Axle ...   
109858  Data May Be Invalid Relative Speed; Rear Axle ...   
130411  Data May Be Invalid Relative Speed; Rear Axle ...   
167370  Data May Be Invalid Relative Speed; Rear Axle ...   
190360  Abnormal Update Rate Engine Instantaneous Fuel...   
275289                                                NaN   
277270                                                NaN   
278386                                                NaN   
282142  Incorrect Data J1939 Network #1 Primary Vehicl...   
533850  Abnormal Update Rate Engine Instantaneous Fuel...   

       ecu_software_version ecu_serial_number ecu_model ecu_make  ecu_source  \
17519       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
57330       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
98748       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
109858      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
130411      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
167370      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
190360              unknown           unknown   unknown  unknown          49   
275289      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
277270      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
278386      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
282142              unknown           unknown   unknown  unknown          11   
533850              unknown           unknown   unknown  unknown          49   

         spn  fmi  active  active_transition_count equipment_id  mct_number  \
17519    907   19    True                       38          305   105362919   
57330    907   19    True                       63          305   105362919   
98748    907   19    True                       86          305   105362919   
109858   907   19    True                       87          305   105362919   
130411   907   19    True                       95          305   105362919   
167370   907   19    True                      106          305   105362919   
190360   184    9    True                      126          305   105362919   
275289  5396    1    True                       78          305   105362919   
277270  5396    1    True                       81          305   105362919   
278386  5396    1    True                       85          305   105362919   
282142   639    2    True                      127          305   105362919   
533850   184    9    True                      126          303   105435252   

         latitude  longitude  location_timestamp accelerator_pedal  \
17519   34.459583 -84.919537 2015-04-10 15:02:32               100   
57330   34.402222 -84.919398 2015-05-19 21:21:34               100   
98748   34.628472 -84.868333 2015-06-27 06:02:21              51,2   
109858  34.786064 -84.824027 2015-07-07 05:32:19               100   
130411  34.803842 -85.012731 2015-07-22 22:34:36              16,8   
167370  33.394120 -84.786435 2015-08-17 09:37:13             

More commas to replace

##### Engine Time (LTD)

In [49]:
faults_diagnostic['engine_time_ltd'].value_counts()

0           584
8890.55     253
4307.4       82
4277.55      76
4422.35      71
           ... 
12052.95      1
1181.9        1
10182.2       1
1517.65       1
4727.45       1
Name: engine_time_ltd, Length: 201163, dtype: int64

In [50]:
faults_diagnostic[faults_diagnostic['engine_time_ltd'].str.contains(",", regex=False, na=False)]

record_id    ess_id     event_timestamp  \
17519       18402   1823305 2015-04-10 14:31:59   
57330       59400   2517346 2015-05-19 20:48:51   
85617       88003   2964433 2015-06-15 15:01:09   
87186       89572   2990820 2015-06-16 22:43:01   
98748      101134   3180493 2015-06-27 05:46:01   
106027     108413   3274660 2015-07-02 12:00:22   
109858     112244   3326811 2015-07-07 05:31:43   
118105     120491   3443634 2015-07-13 14:24:52   
130411     132797   3623872 2015-07-22 22:17:35   
143445     145831   3788565 2015-07-31 09:59:49   
143446     145832   3788566 2015-07-31 09:59:49   
167370     169756   4093848 2015-08-17 09:36:38   
190360     192746   4441365 2015-09-04 04:46:23   
274690     278510   5786929 2015-11-15 08:13:42   
275284     279104   5795238 2015-11-16 02:53:57   
275285     279105   5795239 2015-11-16 02:53:57   
275289     279109   5795315 2015-11-16 03:04:21   
277270     281776   5827502 2015-11-17 08:23:45   
278386     282892   5845487 2015-11-18 07:58:38   
282142     286648   5909821 2015-11-20 22:30:20   
461898     471015   9081774 2016-05-09 05:54:14   
503200     514149  10052557 2016-06-28 10:30:05   
533850     545727  10702825 2016-08-01 02:09:42   

                                        event_description  \
17519   Data May Be Invalid Relative Speed; Rear Axle ...   
57330   Data May Be Invalid Relative Speed; Rear Axle ...   
85617   Data May Be Invalid Relative Speed; Rear Axle ...   
87186   Data May Be Invalid Relative Speed; Rear Axle ...   
98748   Data May Be Invalid Relative Speed; Rear Axle ...   
106027                High (Severity Medium) Engine Speed   
109858  Data May Be Invalid Relative Speed; Rear Axle ...   
118105  Data May Be Invalid Relative Speed; Rear Axle ...   
130411  Data May Be Invalid Relative Speed; Rear Axle ...   
143445     Special Instructions System Diagnostic Code #1   
143446    Not Reporting Data Wheel Sensor ABS Axle 2 Left   
167370  Data May Be Invalid Relative Speed; Rear Axle ...   
190360  Abnormal Update Rate Engine Instantaneous Fuel...   
274690                                                NaN   
275284  Data May Be Invalid Relative Speed; Rear Axle ...   
275285   Low (Severity High) Wheel Sensor ABS Axle 2 Left   
275289                                                NaN   
277270                                                NaN   
278386                                                NaN   
282142  Incorrect Data J1939 Network #1 Primary Vehicl...   
461898  Incorrect Data J1939 Network #1 Primary Vehicl...   
503200  Abnormal Update Rate Engine Instantaneous Fuel...   
533850  Abnormal Update Rate Engine Instantaneous Fuel...   

       ecu_software_version ecu_serial_number ecu_model ecu_make  ecu_source  \
17519       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
57330       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
85617       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
87186       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
98748       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
106027              unknown           unknown   unknown  unknown          49   
109858      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
118105      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
130411      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
143445              unknown           unknown   unknown  unknown          11   
143446              unknown           unknown   unknown  unknown          11   
167370      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
190360              unknown           unknown   unknown  unknown          49   
274690      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
275284      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
275285              unknown           

Replace commas 

##### Fuel Level

In [51]:
faults_diagnostic['fuel_level'].value_counts()

100         42171
60           6898
48           6729
50           6181
67.2         6175
43.2         5777
34.8         5522
53.2         5476
72.4         5426
60.4         5425
50.8         5304
54.4         5069
64           4977
54           4843
40           4763
65.2         4600
54.8         4418
39.2         4191
56.8         4170
61.2         4148
44.8         4144
48.4         4112
51.2         4046
57.2         3993
44.4         3965
50.4         3861
59.2         3761
52.8         3718
38.8         3708
58           3667
0            3520
56.4         3512
62.4         3453
64.4         3450
48.8         3432
44           3409
46.8         3328
72           3305
63.2         3276
49.2         3217
98.4         3190
78.4         3170
60.8         3146
56           3126
82           3110
42.4         3048
65.6         3037
47.2         3031
55.2         2989
46.4         2945
49.6         2885
70.4         2868
42.8         2842
58.4         2836
78           2814
64.8      

More commas to replace

##### Fuel LTD

In [52]:
faults_diagnostic['fuel_ltd'].value_counts()

0                  2923
52878.791304736     250
29786.191379156     109
0.132086026         103
29476.185476134      73
                   ... 
32721.407048928       1
47586.368414968       1
76258.282078788       1
76703.279900382       1
62127.058587178       1
Name: fuel_ltd, Length: 328682, dtype: int64

In [53]:
faults_diagnostic[faults_diagnostic['fuel_ltd'].str.contains(',', regex=False, na=False)]

record_id    ess_id     event_timestamp  \
17519       18402   1823305 2015-04-10 14:31:59   
57330       59400   2517346 2015-05-19 20:48:51   
85617       88003   2964433 2015-06-15 15:01:09   
87186       89572   2990820 2015-06-16 22:43:01   
98748      101134   3180493 2015-06-27 05:46:01   
106027     108413   3274660 2015-07-02 12:00:22   
109858     112244   3326811 2015-07-07 05:31:43   
118105     120491   3443634 2015-07-13 14:24:52   
130411     132797   3623872 2015-07-22 22:17:35   
167370     169756   4093848 2015-08-17 09:36:38   
190360     192746   4441365 2015-09-04 04:46:23   
274690     278510   5786929 2015-11-15 08:13:42   
275284     279104   5795238 2015-11-16 02:53:57   
275285     279105   5795239 2015-11-16 02:53:57   
275289     279109   5795315 2015-11-16 03:04:21   
277270     281776   5827502 2015-11-17 08:23:45   
278386     282892   5845487 2015-11-18 07:58:38   
282142     286648   5909821 2015-11-20 22:30:20   
384770     392271   7596496 2016-02-22 09:58:34   
461898     471015   9081774 2016-05-09 05:54:14   
503200     514149  10052557 2016-06-28 10:30:05   
533850     545727  10702825 2016-08-01 02:09:42   

                                        event_description  \
17519   Data May Be Invalid Relative Speed; Rear Axle ...   
57330   Data May Be Invalid Relative Speed; Rear Axle ...   
85617   Data May Be Invalid Relative Speed; Rear Axle ...   
87186   Data May Be Invalid Relative Speed; Rear Axle ...   
98748   Data May Be Invalid Relative Speed; Rear Axle ...   
106027                High (Severity Medium) Engine Speed   
109858  Data May Be Invalid Relative Speed; Rear Axle ...   
118105  Data May Be Invalid Relative Speed; Rear Axle ...   
130411  Data May Be Invalid Relative Speed; Rear Axle ...   
167370  Data May Be Invalid Relative Speed; Rear Axle ...   
190360  Abnormal Update Rate Engine Instantaneous Fuel...   
274690                                                NaN   
275284  Data May Be Invalid Relative Speed; Rear Axle ...   
275285   Low (Severity High) Wheel Sensor ABS Axle 2 Left   
275289                                                NaN   
277270                                                NaN   
278386                                                NaN   
282142  Incorrect Data J1939 Network #1 Primary Vehicl...   
384770  Incorrect Data J1939 Network #1 Primary Vehicl...   
461898  Incorrect Data J1939 Network #1 Primary Vehicl...   
503200  Abnormal Update Rate Engine Instantaneous Fuel...   
533850  Abnormal Update Rate Engine Instantaneous Fuel...   

       ecu_software_version ecu_serial_number ecu_model ecu_make  ecu_source  \
17519       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
57330       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
85617       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
87186       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
98748       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
106027              unknown           unknown   unknown  unknown          49   
109858      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
118105      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
130411      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
167370      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
190360              unknown           unknown   unknown  unknown          49   
274690      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
275284      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
275285              unknown           unknown   unknown  unknown          11   
275289      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
277270      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
278386      PC4__1284P4C_2*   ________Y048665        MX     PCAR      

More commas

##### Fuel Rate

In [54]:
faults_diagnostic['fuel_rate'].value_counts()

0             97883
0.6075976      6154
0.7528927      5957
0.6340149      5885
0.6604322      5711
0.5679717      5633
0.6868495      5498
0.7132668      5481
0.5415544      5476
0.77931        5348
0.8057272      4396
0.594389       4291
0.5151371      4279
0.5811803      4217
0.7661014      4143
0.6208063      3922
0.6736408      3868
0.739684       3799
0.6472235      3745
0.5547631      3738
0.7000582      3646
0.8321446      3592
0.7264754      3467
0.7925186      3458
0.5283458      3306
0.8585618      3078
0.8849792      2789
0.8189359      2781
0.4887198      2683
0.5019284      2346
0.9246051      2341
0.9510224      2201
0.8453532      2192
1.003857       2188
0.9774396      2183
1.0699         2153
1.096317       2139
1.030274       2086
0.9113964      1959
0.8717705      1908
1.122735       1806
0.8981878      1748
1.149152       1745
0.964231       1705
0.4623025      1693
0.9378137      1673
1.175569       1644
1.215195       1588
0.9906483      1573
1.017066       1542


Commas to replace

##### Fuel Temperature

In [55]:
faults_diagnostic['fuel_temperature'].value_counts()

32       282519
183.2       661
181.4       566
185         449
114.8       393
120.2       363
129.2       360
123.8       353
111.2       350
127.4       350
179.6       349
105.8       346
102.2       336
116.6       326
118.4       320
122         319
96.8        318
109.4       312
98.6        304
100.4       302
125.6       294
93.2        290
89.6        287
113         286
95          285
91.4        278
107.6       275
84.2        269
131         255
80.6        255
104         250
186.8       244
87.8        244
86          241
77          223
132.8       213
82.4        213
134.6       205
78.8        201
136.4       187
75.2        179
73.4        167
140         156
138.2       151
68          149
141.8       149
66.2        144
188.6       143
71.6        139
69.8        137
177.8       136
64.4        111
143.6       104
62.6         99
145.4        98
150.8        97
147.2        94
60.8         92
59           91
46.4         88
149          84
190.4        74
192.2   

Commas

##### IGN Status

In [56]:
faults_diagnostic['ign_status'].value_counts(normalize=True)

True     0.99566
False    0.00434
Name: ign_status, dtype: float64

##### Intake Manifold Temperature

In [57]:
faults_diagnostic['intake_manifold_temperature'].value_counts()

120.2    16720
118.4    16414
102.2    15996
116.6    15896
114.8    15511
100.4    15212
113      14813
98.6     14607
111.2    14472
109.4    13974
96.8     13828
107.6    13494
125.6    13163
105.8    12981
95       12976
123.8    12860
127.4    12839
129.2    12812
122      12780
93.2     12745
131      12626
132.8    12515
104      12426
134.6    12270
136.4    12156
84.2     11825
91.4     11813
138.2    11733
89.6     11110
82.4     10759
87.8     10716
86        9927
80.6      9814
78.8      9100
77        8284
140       8244
141.8     7925
75.2      7640
143.6     7360
73.4      6787
145.4     6626
71.6      6285
66.2      6023
147.2     5866
69.8      5614
68        5333
64.4      5133
149       4753
62.6      4509
60.8      3980
150.8     3923
59        3419
152.6     2979
57.2      2930
55.4      2730
154.4     2342
53.6      2219
51.8      2080
48.2      1803
50        1783
156.2     1679
46.4      1499
44.6      1280
42.8      1185
41        1052
158        931
39.2      

Commas and negative values

##### Lamp Status

In [58]:
faults_diagnostic['lamp_status'].value_counts()

1023     352603
1279     314278
255      301225
17407     99876
18431     28314
2047      19602
2         11667
50175      9341
63487      8220
62463      7860
16639      7620
22527      5738
51199      4373
0          3664
21503      2758
65535      1981
2035       1470
6143       1301
5119        825
17663       627
4351        484
55295       391
9           345
5375        217
18419       216
20735        60
11           33
544          26
511          23
50431        10
16895         5
22515         4
28436         4
11801         2
617           1
6131          1
65523         1
Name: lamp_status, dtype: int64

Nothing to clean!

##### Parking Brake

In [59]:
faults_diagnostic['parking_brake'].value_counts(normalize=True)

False    0.667462
True     0.332538
Name: parking_brake, dtype: float64

30% of the dataset includes trucks that are parked. Will we want to use this?

##### Service Distance

In [60]:
faults_diagnostic['service_distance'].value_counts()

-10774.576171875     8
16500.51171875       4
-99813.9609375       3
-23487.83203125      3
-23770.5546875       2
17370.431640625      2
-13235.2060546875    2
-21334.779296875     2
37953.3515625        1
42203.53125          1
5635.8369140625      1
37043.04296875       1
-22506.064453125     1
39811.25390625       1
38546.76171875       1
826.423706054688     1
-23528.220703125     1
-9214.9345703125     1
-8093.35986328125    1
-20784.8671875       1
-17808.498046875     1
-23388.412109375     1
-16332.7421875       1
37350.62109375       1
38105.58984375       1
-11678.671875        1
-21766.6328125       1
-2951.51318359375    1
37002.65625          1
39668.3359375        1
40019.41015625       1
-21760.419921875     1
38456.6640625        1
37810.4375           1
4737.95556640625     1
-17308.294921875     1
-21107.978515625     1
-16255.0703125       1
-15742.439453125     1
-23646.28125         1
-22509.171875        1
-10215.3427734375    1
37213.921875         1
-21751.0976

There are very few records within this column (~400 out of a 1.18 million), so we're safe to drop it

In [61]:
faults_diagnostic = faults_diagnostic.drop(columns='service_distance')

##### Speed

In [62]:
faults_diagnostic['speed'].value_counts()

0              199216
0.009708925      1613
0.01941785       1025
66.88478          837
66.83624          795
                ...  
64.08133            1
73.28297            1
3.720945            1
77.09857            1
31.04429            1
Name: speed, Length: 13533, dtype: int64

In [63]:
faults_diagnostic[faults_diagnostic['speed'].str.contains(',', regex=False, na=False)]

record_id    ess_id     event_timestamp  \
17519       18402   1823305 2015-04-10 14:31:59   
57330       59400   2517346 2015-05-19 20:48:51   
85617       88003   2964433 2015-06-15 15:01:09   
87186       89572   2990820 2015-06-16 22:43:01   
98748      101134   3180493 2015-06-27 05:46:01   
106027     108413   3274660 2015-07-02 12:00:22   
109858     112244   3326811 2015-07-07 05:31:43   
118105     120491   3443634 2015-07-13 14:24:52   
130411     132797   3623872 2015-07-22 22:17:35   
167370     169756   4093848 2015-08-17 09:36:38   
190360     192746   4441365 2015-09-04 04:46:23   
275284     279104   5795238 2015-11-16 02:53:57   
275285     279105   5795239 2015-11-16 02:53:57   
275289     279109   5795315 2015-11-16 03:04:21   
282142     286648   5909821 2015-11-20 22:30:20   
384770     392271   7596496 2016-02-22 09:58:34   
461898     471015   9081774 2016-05-09 05:54:14   
503200     514149  10052557 2016-06-28 10:30:05   
533850     545727  10702825 2016-08-01 02:09:42   

                                        event_description  \
17519   Data May Be Invalid Relative Speed; Rear Axle ...   
57330   Data May Be Invalid Relative Speed; Rear Axle ...   
85617   Data May Be Invalid Relative Speed; Rear Axle ...   
87186   Data May Be Invalid Relative Speed; Rear Axle ...   
98748   Data May Be Invalid Relative Speed; Rear Axle ...   
106027                High (Severity Medium) Engine Speed   
109858  Data May Be Invalid Relative Speed; Rear Axle ...   
118105  Data May Be Invalid Relative Speed; Rear Axle ...   
130411  Data May Be Invalid Relative Speed; Rear Axle ...   
167370  Data May Be Invalid Relative Speed; Rear Axle ...   
190360  Abnormal Update Rate Engine Instantaneous Fuel...   
275284  Data May Be Invalid Relative Speed; Rear Axle ...   
275285   Low (Severity High) Wheel Sensor ABS Axle 2 Left   
275289                                                NaN   
282142  Incorrect Data J1939 Network #1 Primary Vehicl...   
384770  Incorrect Data J1939 Network #1 Primary Vehicl...   
461898  Incorrect Data J1939 Network #1 Primary Vehicl...   
503200  Abnormal Update Rate Engine Instantaneous Fuel...   
533850  Abnormal Update Rate Engine Instantaneous Fuel...   

       ecu_software_version ecu_serial_number ecu_model ecu_make  ecu_source  \
17519       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
57330       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
85617       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
87186       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
98748       PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
106027              unknown           unknown   unknown  unknown          49   
109858      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
118105      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
130411      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
167370      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
190360              unknown           unknown   unknown  unknown          49   
275284      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
275285              unknown           unknown   unknown  unknown          11   
275289      PC4__1284P4C_2*   ________Y048665        MX     PCAR           0   
282142              unknown           unknown   unknown  unknown          11   
384770              unknown           unknown   unknown  unknown          11   
461898  BB41103*   BB41104*        S041400607  EC60-adv    BNDWS          11   
503200              unknown           unknown   unknown  unknown          49   
533850              unknown           unknown   unknown  unknown          49   

         spn  fmi  active  active_transition_count equipment_id  mct_number  \
17519    907   19    True                       38          305   105362919   
57330  

Commas to replace

##### Switched Battery Voltage

In [64]:
faults_diagnostic['switched_battery_voltage'].value_counts()

3276.75     102907
14            1375
13.95         1097
13.9           962
13.85          858
14.05          803
13.8           521
13.75          441
14.1           441
13.7           322
13.6           234
13.65          227
12.5           178
14.15          157
13.55          151
12.55          142
13.5           140
12.6           131
12.75          126
12.7           126
12.8           126
12.4           115
12.25          113
12.45          113
13.05          108
13.45          102
13.25          101
12.65          100
12.85          100
13.15           99
13.1            96
13.35           95
12.9            94
12.35           93
13.3            89
13.2            86
12.95           85
13              85
12.3            84
13.4            73
12.2            62
12.15           48
12              47
11.8            47
12.1            45
12.05           45
11.85           40
14.2            34
11.7            28
11.95           25
11.75           17
11.9            16
3276,75     

Commas to replace

##### Throttle

In [65]:
faults_diagnostic['throttle'].value_counts()

100         285394
0           113358
38.4           226
37.6           225
38.8           224
40.8           214
36.4           210
38             207
39.2           202
35.2           201
33.6           199
34.4           196
34.8           195
42.8           193
34             192
37.2           192
0.4            191
28.8           189
36             187
32.8           187
39.6           184
41.2           182
36.8           182
42             182
44             178
32.4           178
33.2           177
40.4           176
44.8           175
31.6           172
40             170
41.6           169
30             167
42.4           166
30.8           165
43.6           163
35.6           162
28.4           160
32             160
46.4           158
99.6           154
46             152
44.4           152
43.2           151
45.6           150
31.2           150
30.4           150
29.2           143
27.6           141
45.2           141
29.6           140
99.2           134
47.6        

Commas

##### Turbo Boost Pressure

In [66]:
faults_diagnostic['turbo_boost_pressure'].value_counts()

0.29     94718
0        66474
0.58     52541
0.87     29056
1.16     24923
1.45     22229
1.74     19455
2.03     16948
2.32     14241
2.61     11765
2.9      10017
3.19      8799
3.48      8026
3.77      7356
4.06      6582
4.35      6018
4.64      5406
4.93      4856
5.22      4507
5.51      4167
5.8       3904
6.09      3767
6.38      3608
6.67      3333
6.96      3277
7.25      3255
7.54      3035
7.83      2878
8.12      2837
8.41      2718
8.7       2661
9.28      2599
8.99      2583
9.57      2419
10.15     2325
9.86      2294
10.73     2265
25.52     2236
11.02     2201
10.44     2185
26.68     2184
26.1      2168
26.39     2166
25.81     2162
26.97     2117
25.23     2087
11.31     2085
12.47     2081
11.89     2061
27.26     2049
11.6      2018
12.18     2005
27.55     1995
24.94     1929
13.34     1917
12.76     1893
13.63     1873
24.65     1862
13.05     1844
27.84     1834
28.13     1820
24.36     1809
14.21     1745
13.92     1729
28.42     1699
14.5      1682
14.79     

Commas

## Replacing commas and changing to floats

In [67]:
comma_columns = [
    'accelerator_pedal',
    'barometric_pressure',
    'distance_ltd',
     'engine_coolant_temperature',
     'engine_oil_pressure',
     'engine_oil_temperature',
     'engine_rpm',
     'engine_time_ltd',
     'fuel_level',
     'fuel_ltd',
     'fuel_rate',
     'fuel_temperature',
     'intake_manifold_temperature',
     'speed',
     'switched_battery_voltage',
     'throttle',
     'turbo_boost_pressure']

for column in comma_columns:
    faults_diagnostic[column] = faults_diagnostic[column].str.replace(',', '.')
    faults_diagnostic[column] = faults_diagnostic[column].astype('float64')

## Use forward fill and backfill to reduce null values

In [68]:
def input_fill(df):
    return df.sort_values(by='event_timestamp').fillna(method='ffill').fillna(method='bfill')

In [69]:
faults_diagnostic = faults_diagnostic.groupby('equipment_id').apply(input_fill)

In [70]:
faults_diagnostic.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 1185166 entries, ('1327', 4967) to ('R1764', 4999)
Data columns (total 41 columns):
 #   Column                       Non-Null Count    Dtype         
---  ------                       --------------    -----         
 0   record_id                    1185166 non-null  int64         
 1   ess_id                       1185166 non-null  int64         
 2   event_timestamp              1185166 non-null  datetime64[ns]
 3   event_description            1185166 non-null  object        
 4   ecu_software_version         1185003 non-null  object        
 5   ecu_serial_number            1185003 non-null  object        
 6   ecu_model                    1185003 non-null  object        
 7   ecu_make                     1185003 non-null  object        
 8   ecu_source                   1185166 non-null  int64         
 9   spn                          1185166 non-null  int64         
 10  fmi                          1185166 non-null  int64     

## Data Types

In [71]:
string_columns = [
    'record_id',
    'ess_id',
    'ecu_source',
    'spn',
    'fmi',
    'mct_number'
]

integers = [
    'engine_load'
]

bools = [
    'cruise_control_active',
    'ign_status',
    'parking_brake'
]

floats = [
    'cruise_control_set_speed',
    'service_distance'
]

In [72]:
for column in string_columns:
    faults_diagnostic[column] = faults_diagnostic[column].apply(lambda x: str(x))
    
for column in integers:
    faults_diagnostic[column] = pd.to_numeric(faults_diagnostic[column], errors='coerce')
    
for column in bools:
    faults_diagnostic[column] = faults_diagnostic[column].apply(lambda x: bool(x))
    
for column in floats:
    faults_diagnostic[column] = faults_diagnostic[column].apply(lambda x: float(x))

KeyError: 'service_distance'

In [ ]:
faults_diagnostic.info()

#### Removing service station records using Lat/Long

In [ ]:
service_centers = faults_diagnostic[ 
        ((faults_diagnostic['latitude'].between(36.05, 36.07, inclusive=True)) & (faults_diagnostic['longitude'].between(-86.44, -86.42, inclusive=True))) |
        ((faults_diagnostic['latitude'].between(35.57, 35.59, inclusive=True)) & (faults_diagnostic['longitude'].between(-86.45, -86.43, inclusive=True))) |
        ((faults_diagnostic['latitude'].between(36.18, 36.21, inclusive=True)) & (faults_diagnostic['longitude'].between(-83.18, -83.16, inclusive=True)))
        ].index

In [ ]:
faults_diagnostic = faults_diagnostic.drop(service_centers, axis=0)

In [ ]:
faults_diagnostic.shape

#### Ungrouping by Equipment ID

This will ensure our cleaned dataset will match our unseen data

In [ ]:
#equipment_id became and index and the column was retained. We drop it first to avoid an error
#after resetting the index, a 'level_1' column is created, so that is dropped as well

faults_diagnostic = faults_diagnostic.drop(columns =['equipment_id']).reset_index().drop(columns=['level_1'])

#### Dropping Active = False

In [ ]:
faults_diagnostic = faults_diagnostic[faults_diagnostic['active'] == True]

## Creating Target Column

In [ ]:
faults_diagnostic['5246_derate'] = faults_diagnostic['spn'].apply(lambda x: 'True' if x == '5246' else 'False') 

In [ ]:
faults_diagnostic[faults_diagnostic['5246_derate'] == 'True']

In [ ]:
faults_diagnostic['1569_derate'] = faults_diagnostic['spn'].apply(lambda x: 'True' if x == '1569' else 'False')

In [ ]:
faults_diagnostic[faults_diagnostic['1569_derate'] == 'True']

## Export as CSV

In [ ]:
faults_diagnostic.to_csv('../data/cleaned_faults_diagnostic.csv', index=False)